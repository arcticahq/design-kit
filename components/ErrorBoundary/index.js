import React from "react"
import PropTypes from "prop-types"
import Button from "../Button"

class ErrorBoundary extends React.Component {
  static handleReload() {
    window.location.reload()
  }

  state = { hasError: false }

  componentDidCatch(error, errorInfo) {
    this.setState({ hasError: true })
    Sentry.configureScope(scope => {
      Object.keys(errorInfo).forEach(key => {
        scope.setExtra(key, errorInfo[key])
      })
    })
    Sentry.captureException(error)
  }

  render() {
    const { children } = this.props
    const { hasError } = this.state

    if (hasError) {
      return (
        <>
          <h2 className="text-center text-white font-bold mt-5" role="alert">
            Something went wrong
          </h2>
          <p className="text-center mt-5">
            <Button primary onClick={e => ErrorBoundary.handleReload(e)}>
              Reload
            </Button>
          </p>
        </>
      )
    }
    return children
  }
}

ErrorBoundary.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.array,
    PropTypes.object,
    PropTypes.element
  ]).isRequired
}

export default ErrorBoundary
