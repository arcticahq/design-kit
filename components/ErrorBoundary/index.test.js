import React from "react"
import { shallow } from "enzyme"
import ErrorBoundary from "."

describe(`ErrorBoundary`, () => {
  let props
  let mountedComponent
  const wrapper = () => {
    if (!mountedComponent) {
      mountedComponent = shallow(
        <ErrorBoundary {...props}>{props.children}</ErrorBoundary>
      )
    }
    return mountedComponent
  }

  beforeEach(() => {
    props = {
      children: <div>Hello</div>
    }
    mountedComponent = undefined
  })

  it(`renders a ErrorBoundary correctly`, () => {
    wrapper().setState({ hasError: true })
    expect(wrapper()).toMatchSnapshot()
  })
})
