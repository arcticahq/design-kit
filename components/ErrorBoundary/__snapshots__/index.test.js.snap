// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorBoundary renders a ErrorBoundary correctly 1`] = `
<Fragment>
  <h2
    className="text-center text-white font-bold mt-5"
    role="alert"
  >
    Something went wrong
  </h2>
  <p
    className="text-center mt-5"
  >
    <Button
      className={null}
      disabled={false}
      noink={false}
      onClick={[Function]}
      primary={true}
      transparent={false}
      type="button"
    >
      Reload
    </Button>
  </p>
</Fragment>
`;
